### ISSUE TITLE
map_create_pixelart tool issue board

### ISSUE THREAD

**Issue by AssassinTee** [ddnet/ddnet#8345]
I am currently working with the [`map_create_pixelart`](https://github.com/ddnet/ddnet/pull/5766) tool, which converts a pixelart (png) into quads and wanted to share issues and missing features:

### No download available
While this tool is part of ddnet and can be build with `make --build build --target tools`, there is no download available, neither on the ddnet website (also tested nightly) or here on github. This issue is big, because most people aren't able to compile for themself.

### Path length
 The path of the map/image isn't long enough, I tried the following:
```
./ddnet-tools/map_create_pixelart.exe pixelart/img_masked/ice_mountain_12.png 1 ~/AppData/Roaming/Teeworlds/maps/JumpKingDitherTest.map 0 0 0 0 4 output-map.map 1 0

2024-05-10 17:26:12 I map_create_pixelart: image_file='pixelart/img_masked/ice_mountain_12.png'; image_pixelsize='1px'; input_map='output-map.map'; layergroup_id='#0'; layer_id='#0'; pos_x='#0px'; pos_y='0px'; quad_pixelsize='4px'; output_map='pixelart/img_masked/ice_mountain_12.png'; optimize='1'; centralize='0'
2024-05-10 17:26:12 I storage: added path '$CURRENTDIR' ('C:\Users\<USER>\Desktop\workspace\JumpKingScript')
2024-05-10 17:26:12 I datafile: could not open 'C:/Users/<USER>/AppData/Roaming/Teeworlds/maps/JumpKingDitherTe'
2024-05-10 17:26:12 I map_create_pixelart: ERROR: unable to open map 'C:/Users/<USER>/AppData/Roaming/Teeworlds/maps/JumpKingDitherTe'
```
As you can see, the input map is cut off. My guess, is that the number of bytes allocated for the path aren't big enough. I worked around this by copying the map in another location (with shorter path), then use the tool, then copy it back.

### wrong variable repetition

As you can see above, the variables, which I entered into the script, didn't get echoed back right:
```
2024-05-10 17:26:12 I map_create_pixelart: image_file='pixelart/img_masked/ice_mountain_12.png'; image_pixelsize='1px'; input_map='output-map.map'; layergroup_id='#0'; layer_id='#0'; pos_x='#0px'; pos_y='0px'; quad_pixelsize='4px'; output_map='pixelart/img_masked/ice_mountain_12.png'; optimize='1'; centralize='0'
```
`input_map` is wrong
`output_map` is echoed wrong

### input map equal to output map
when input map and output map are equal, it either ends with a segmentation fault or with an empty 0 byte map. So I recommend to always do backups before doing CLI-mapping operations

### Add group and layer automatically
instead of needing a group and a layer, just add a group with a layer. You can still move it afterwards

### Default values
this tool has 12 values and needs all for them in order to run. Add default values to most of them, all you really need is input_map, input_image, quad_pixelsize and potentially output_map. Everything else can be assumed:

No value for
```
layergroup_id
layer_id
```
assume create new instead

you can always move a group
```
pos_x = 0
pos_y = 0
```

```
optimize=1
centralize=0
```

### input_map eq output_map
allow input_map and output_map to be the same in order to allow in-situ operations

### Add Quadart

In the editor, you can press `Tools->Add Tileart`, why not add Quadart and call this tool?

## Workaround
I worked around most of this issues with a custom bash script, ask me on discord if you want it

---

I might do a PR myself in the future if I find time

**Comment by BlaiZephyr** [comment]
this is insanely detailed and very interesting :1st_place_medal:

**Comment by AssassinTee** [comment]
@furo321 Thank you very much! This was very fast

**Comment by BlaiZephyr** [comment]
> (missing) features
remain unimplemented which for me still look like QoL we'd want - maybe make it a new issue or reopen this with a new title?

**Comment by AssassinTee** [comment]
> > (missing) features
> 
> remain unimplemented which for me still look like QoL we'd want - maybe make it a new issue or reopen this with a new title?

Reopened as feature request

**Comment by sctt** [comment]
thank you for improving my tool :)
i just want to answer this question:

> Add Quadart
> In the editor, you can press Tools->Add Tileart, why not add Quadart and call this tool?

i agree that this is something that should definitely be done, not only for this tool, but also for stuff like `map_replace_area` that is very powerful when you need to copy all tiles and quads from a place to another (considering also parallax, offsets, etc.).

the reason why those are currently CLI tools and not editor features is very simple: i'm good in writing algorithms, but i have not enough knowledge to implement them inside the editor context (it's not you just need to "call the tool": you have to port the logic directly inside the editor code).

i hope some good soul who has the required editor coding experience will be nice enough to perform this porting (especially for `map_replace_area`) as many mappers asked me about this but i currently have no time to get deep inside the editor logic in order to satisfy the request.

i'll keep crossing fingers hoping to see this in the future :)

thanks to everybody!

**Comment by AssassinTee** [comment]
Some last words:

> I might do a PR myself in the future if I find time

I found time after a year 🥳 

I actually implemented **all** the features from my list by integrating it as a tool. This removed the need for loading and saving a map at all and making it trivial to add a group and a layer directly into a map. This also removed the need for StartX/StartY and we now have only 4 configuration values left:
- ImagePixelSize
- QuadPixelSize
- Centralize
- Optimize

Also it allowed to set details and a clip automatically, which can greatly improve FPS for some weaker machines.

The algorithm is still the original from @sctt , but rewritten to follow the repository style and to increase performance at some important corner stones.

I think it turned out really nice

**Comment by sctt** [comment]
@AssassinTee thank you for integrating this to the map editor.

i just want to add a note: you included in the new code the `FindSuperPixelSize` function, which is currently called nowhere, as it was part of the `GetImagePixelSize` routine which wasn't included in your changes, so i guess it can be considered useless so you can feel free to remove that.... unless you decide to actually use that algorithm and take advantage of its feature.

those routines were needed to automatically deduct the `ImagePixelSize` parameter, of course users can decide to define that by their own but that could lead to a messed up result and/or useless pixels, especially if the optimize feature is turned off (sometimes you dont want to have optimization on as you might want to apply different envs colors or offsets to specific pixels in the grid).

you could call the `ImagePixelSize` function everytime the user opens the "Configure Quadart" box so it will automatically propose a value in the "Image pixelsize" field, or use some other caching strategy to avoid useless multiple calls of this routine, maybe you can let the user choose with a button "Automatically caculate Image pixelsize".

of course this is your decision, i'm writing this only to let you know about a feature you might have missed.

best regards.

**Comment by AssassinTee** [comment]
@sctt usually a pixelart has a pixelsize of 1 pixel. Using a higher value pixelizes your image. So this is a buildin feature. For example I can load in the mona-lisa and just use every 20th pixel getting a pixelized mona lisa.
I see _little_ value in calculating this automatically, as it's just more code to maintain. On the other side I also see it's benefits and may add a followup to add it into the editor. I'd see greater benefit in improving the optimize functionality, so that it doesn't lead to useless pixels or directly recalculate img_pixel_size and quad_pixel_size for this cases (like set img pixel size to 1, but the true img pixel size is 2, and then recalculate quad_pixel_size)

**Comment by sctt** [comment]
@AssassinTee thank you for your attention, i just want to add a couple of notes just to clarify better the exact use case i was referring to

> [@sctt](https://github.com/sctt) usually a pixelart has a pixelsize of 1 pixel. Using a higher value pixelizes your image. So this is a buildin feature. For example I can load in the mona-lisa and just use every 20th pixel getting a pixelized mona lisa. 

the mona lisa you are citing its not a pixel art image, its a normal image you are transforming into a pixelart thanks to the logic you described: on that use case calculating the image pixel size its quite useless, as it will practically always be 1 pixel, the user wants to change it in order to make the final result more or less defined.

the use case i was referring to is quite different, its when users have a real pixel art image (like the ones you can create with tools like [this](https://www.pixilart.com/draw)), and want to import it into their map. on those kind of images, the pixelart pixel size will almost never be 1 pixel, but a cluster of pixels users don't really want to change, they only want to know the value in order to have the exact number of quads for the exact number of pixelart pixels in the image, take this example:

![Image](https://github.com/user-attachments/assets/753ae077-b632-49c4-bede-d4b88ba2f281)

this is a real pixel art (no shades, no hues, just plain pixels), and its "super pixel" size is 32x32 "sub pixels", here users just want a quad for every 32x32 square in the image, they are not interested in any other value that is not 32.
we kinda used this use case a lot in our TeeTown map to make stuff like this:

![Image](https://github.com/user-attachments/assets/481b9711-9b61-49ee-ad7d-2521396033a9)

fun thing is that this use case was the main purpose in mind when coding the map_pixelart tool, the other use case you are referring to is just a cool side effect :)

> I see little value in calculating this automatically

the only advantage is to avoid users (when working on a _real_ pixel art image) to open an image editor software in order to manually measure the image "super pixel" size with a ruler: this is not a big deal, having it calculated by default its just a convenient automation for the user.

> I'd see greater benefit in improving the optimize functionality

improving is always good, but consider that activating the optimize option is already a solution to this problem for the final user, as pixels will be grouped in an efficient way despite the fact that the image pixel size was optimal or not.
in summary: automatically calculating the image pixel size is useful only when dealing with _real_ pixel art image in those cases where the optimize option is not desired (like when you want to apply different parameters to single pixels, such as colors, envs, etc).

> as it's just more code to maintain

i want to repeat myself: im not trying to push you to do anything, i just wanted to clarify what i was talking about.
said that, if you are not interested in the `GetImagePixelSize` feature, consider to remove the `FindSuperPixelSize` function, as it is currently an orphan function called by nobody.

i hope i've been useful,
thank you again.
